### DISCUSSION TITLE
Have any way to virtualize controller layer thereby make two bluetooth server and client communicate mutually in only one bluetooth chip?

### DISCUSSION THREAD

**Discussion by Squirre17** [google/bumble#523] [General]
My current environment is Windows + Linux VM. Is there a way for me to set up two Bluetooth devices to communicate with each other within the Linux VM, using only one host Bluetooth module? I want to debug the host layer of Bluetooth. Is there a way to completely virtualize and shield the lower layer of the controller?

**Comment by zxzxwu** [comment]
This is a question more about BlueZ? Bumble focuses on Host implementation, so I guess we cannot help much here.

I don't know much about BlueZ, but I assume it also interacts with an HCI interface somewhere. If so, you can use project like rootcanal to emulate the transport.

However, if you are just debugging, do you really need to do so? If your host layer is not BlueZ but something else, you may just make its HCI interface configurable to work with emulation layers.

**Reply by Squirre17** [reply]
I want to learn how the host layer works, both bluez and bumble work for me, bumble is written by python so it's easier to debug, so I am thinking about how to build the underlying vHCI so that both the server and the client can run on bumble, but I can't find the right vHCI project so far

**Comment by barbibulle** [comment]
If I understand correctly what you're looking for, this should be possible, by using a combination of the Bumble HCI bridge and the RootCanal virtual controller. Rootcanal (https://github.com/google/rootcanal) is a virtual bluetooth controller implementation. You can instantiate a server that exposes virtual HCI interfaces over TCP sockets, and all the host stacks that connect to those HCI interfaces can then communicate with each other.
The Linux BlueZ stack can't use RootCanal directly, but you can use the Bumble HCI bridge application (https://google.github.io/bumble/apps_and_tools/hci_bridge.html) to bridge the Linux VHCI device to a RootCanal controller.
Does that look like what you have in mind?

**Reply by Squirre17** [reply]
Seems like rootcanal is what I'm looking for, thx. But I have a little question about this. Are there any constraints on building rootcanal on x86 Ubuntu? I downloaded and installed it's whl file, but I was missing part of the c library.
```shell
Traceback (most recent call last):
  File "/home/squ/Projects/learning-projs/rootcanal/rootcanal/my.py", line 2, in <module>
    from rootcanal.controller import Controller, Any
  File "/home/squ/.pyenv/versions/3.11.8/lib/python3.11/site-packages/rootcanal/controller.py", line 32, in <module>
    rootcanal = cdll.LoadLibrary(librootcanal_ffi_path)
                ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/squ/.pyenv/versions/3.11.8/lib/python3.11/ctypes/__init__.py", line 454, in LoadLibrary
    return self._dlltype(name)
           ^^^^^^^^^^^^^^^^^^^
  File "/home/squ/.pyenv/versions/3.11.8/lib/python3.11/ctypes/__init__.py", line 376, in __init__
    self._handle = _dlopen(self._name, mode)
                   ^^^^^^^^^^^^^^^^^^^^^^^^^
OSError: libcrypto.so.3: cannot open shared object file: No such file or directory
```
google tell my the `libcrypto.so.3` is android dedicated.

**Reply by barbibulle** [reply]
Rootcanal should build and run on x86 ubuntu. 
Also, note that if your goal is to study how a Bluetooth stack works, going though the Bumble host stack python implementation, you can just just two bumble stacks talking to each other via Rootcanal, you don't need BlueZ and a bridge to vhci.
You simply run Rootcanal, which will accept HCI connections over TCP. With bumble, you use `tcp-client:localhost:<port>` as the transport name for the RootCanal HCI interface (you can connect as many Bumble instance to that, they will each get their own RootCanal controller). The default port for RootCanal over TCP is 6402.

**Reply by Lerr1uqs** [reply]
> Rootcanal should build and run on x86 ubuntu. Also, note that if your goal is to study how a Bluetooth stack works, going though the Bumble host stack python implementation, you can just just two bumble stacks talking to each other via Rootcanal, you don't need BlueZ and a bridge to vhci. You simply run Rootcanal, which will accept HCI connections over TCP. With bumble, you use `tcp-client:localhost:<port>` as the transport name for the RootCanal HCI interface (you can connect as many Bumble instance to that, they will each get their own RootCanal controller). The default port for RootCanal over TCP is 6402.

Hi, I also want to learn how bluetooth protocol host work and found that such awesome project, I have a bluetooth chip and a extra bt adapter. How can I make two gatt server and client communicate each other on the two bt hci. I will very appreciate if you can give me specific command.(I don't know what transport I should select in the gatt example......)
```shell
hcitool dev                                                                                                                                           0 [20:03:48]
Devices:
        hci1    00:E0:4C:32:81:A0
        hci0    00:50:56:F8:B9:29
```

**Reply by Lerr1uqs** [reply]
> > Rootcanal should build and run on x86 ubuntu. Also, note that if your goal is to study how a Bluetooth stack works, going though the Bumble host stack python implementation, you can just just two bumble stacks talking to each other via Rootcanal, you don't need BlueZ and a bridge to vhci. You simply run Rootcanal, which will accept HCI connections over TCP. With bumble, you use `tcp-client:localhost:<port>` as the transport name for the RootCanal HCI interface (you can connect as many Bumble instance to that, they will each get their own RootCanal controller). The default port for RootCanal over TCP is 6402.
> 
> Hi, I also want to learn how bluetooth protocol host work and found that such awesome project, I have a bluetooth chip and a extra bt adapter. How can I make two gatt server and client communicate each other on the two bt hci. I will very appreciate if you can give me specific command.(I don't know what transport I should select in the gatt example......)
> 
> ```shell
> hcitool dev                                                                                                                                           0 [20:03:48]
> Devices:
>         hci1    00:E0:4C:32:81:A0
>         hci0    00:50:56:F8:B9:29
> ```

oh I got it. I always thought that gatt's server and client needed to create two python scripts and pty, and then bridge the two pty's in some form, but in reality I only needed to use LocalLink in one script to do all the interaction without any bt adaptor

**Reply by zxzxwu** [reply]
Yes, for Host study purpose, LocalLink or Rootcanal should be enough. If you want to run Bumble on physical controllers, you need to figure out what's their transport (usually usb, you could check with `apps/usb_probe.py`), and use example scripts provided in `examples/` directory to run GATT client and server on them.
